File: backend-java/src/main/java/com/example/faceattendance/service/AuthService.java
package com.example.faceattendance.service;

import com.example.faceattendance.dto.auth.ChangePasswordRequest;
import com.example.faceattendance.dto.auth.ForgotPasswordRequest;
import com.example.faceattendance.dto.auth.LoginOtpRequiredResponse;
import com.example.faceattendance.dto.auth.LoginRequest;
import com.example.faceattendance.dto.auth.LoginResponse;
import com.example.faceattendance.dto.auth.PasswordActionResponse;
import com.example.faceattendance.dto.auth.ResendForgotPasswordOtpRequest;
import com.example.faceattendance.dto.auth.ResendLoginOtpRequest;
import com.example.faceattendance.dto.auth.VerifyForgotPasswordRequest;
import com.example.faceattendance.dto.auth.VerifyLoginOtpRequest;

public interface AuthService {

    /**
     * Verifies username/password only. On success, triggers a login OTP challenge
     * (stored in Redis + emailed) and returns a non-JWT response. Does NOT issue a token.
     */
    LoginResponse login(LoginRequest request);

    /**
     * Verifies the login OTP for a pending login challenge created by {@link #login}.
     * Only on success is the existing JWT generated and returned.
     */
    LoginResponse verifyLoginOtp(VerifyLoginOtpRequest request);

    /**
     * Resends a login OTP for an existing pending login challenge, subject to cooldown.
     */
    LoginOtpRequiredResponse resendLoginOtp(ResendLoginOtpRequest request);

    PasswordActionResponse requestPasswordReset(ForgotPasswordRequest request);
    PasswordActionResponse verifyPasswordReset(VerifyForgotPasswordRequest request);
    PasswordActionResponse resendPasswordResetOtp(ResendForgotPasswordOtpRequest request);
    PasswordActionResponse changePassword(String username, ChangePasswordRequest request);
}

File: frontend-UI/src/services/authService.js
import api from './api';

const authService = {
  // Step 1:
  // Verify username + password and trigger email OTP.
  login: async (username, password) => {
    const response = await api.post('/api/auth/login', {
      username,
      password,
    });

    return response.data;
  },

  // Step 2:
  // Verify OTP and receive the final JWT.
  verifyLoginOtp: async (username, otp) => {
    const response = await api.post('/api/auth/verify-otp', {
      username,
      otp,
    });

    return response.data;
  },

  // Resend OTP for an existing pending login.
  resendLoginOtp: async (username) => {
    const response = await api.post('/api/auth/resend-otp', {
      username,
    });

    return response.data;
  },

  /**
   * Starts the username-based forgot-password flow.
   * The backend resolves the user's email and sends the OTP.
   * The frontend does not ask the user for an email address.
   */
  requestPasswordReset: async (username) => {
    const response = await api.post(
      '/api/auth/forgot-password',
      { username }
    );

    return response.data;
  },

  verifyPasswordReset: async ({
    username,
    otp,
    newPassword,
    confirmPassword,
  }) => {
    const response = await api.post(
      '/api/auth/forgot-password/verify',
      {
        username,
        otp,
        newPassword,
        confirmPassword,
      }
    );

    return response.data;
  },

  resendPasswordResetOtp: async (username) => {
    const response = await api.post(
      '/api/auth/forgot-password/resend-otp',
      { username }
    );

    return response.data;
  },

  changePassword: async ({
    currentPassword,
    newPassword,
    confirmPassword,
  }) => {
    const response = await api.post(
      '/api/auth/change-password',
      {
        currentPassword,
        newPassword,
        confirmPassword,
      }
    );

    return response.data;
  },
};

export default authService;